importers/rules.go

```diff
@@ -0,0 +1,239 @@
+
+	for _, file := range files {
+		ruleGroups, errs := rulefmt.ParseFile(file)
+		if errs != nil {
+			return errs
+		}
+
+		for _, ruleGroup := range ruleGroups.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rulez := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := parser.ParseExpr(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, file)}
+				}
+
+				rulez = append(rulez, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[file+";"+ruleGroup.Name] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     file,
+				Interval: itv,
+				Rules:    rulez,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return errs
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores in new blocks
+func (importer *RuleImporter) ImportAll(ctx context.Context) error {
+	for _, group := range importer.groups {
+		for _, rule := range group.Rules() {
+			// todo: what happens if there is an error importing one of the rules?
+			importer.ImportRule(ctx, rule)
+		}
+	}
+	return importer.CreateBlocks()
+}
+
+func (importer *RuleImporter) queryFn(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
+	// what to do with warnings?
+	val, _, err := importer.apiClient.Query(ctx, q, t)
+	if err != nil {
+		return promql.Vector{}, err
+	}
+
+	switch val.Type() {
+	case model.ValVector:
+		valVector := val.(model.Vector)
+		return modelToPromqlVector(valVector), nil
+	case model.ValScalar:
+		valScalar := val.(*model.Scalar)
+		return promql.Vector{promql.Sample{
+			Metric: labels.Labels{},
+			Point:  promql.Point{T: int64(valScalar.Timestamp), V: float64(valScalar.Value)},
+		}}, nil
+	default:
+		return nil, errors.New("rule result is wrong type")
+	}
+}
+
+func modelToPromqlVector(modelValue model.Vector) promql.Vector {
+	result := make(promql.Vector, 0, len(modelValue))
+
+	for _, value := range modelValue {
+		labels := make(labels.Labels, 0, len(value.Metric))
+
+		for k, v := range value.Metric {
+			labels = append(labels, plabels.Label{
+				Name:  string(k),
+				Value: string(v),
+			})
+		}
+		sort.Sort(labels)
+
+		result = append(result, promql.Sample{
+			Metric: labels,
+			Point:  promql.Point{T: int64(value.Timestamp), V: float64(value.Value)},
+		})
+	}
+	return result
+}
+
+// ImportRule imports the historical data for a single rule
+func (importer *RuleImporter) ImportRule(ctx context.Context, rule rules.Rule) error {
+	ts, err := parseTime(importer.config.Start)
+	if err != nil {
+		return err
+	}
+	end, err := parseTime(importer.config.End)
+	if err != nil {
+		return err
+	}
+	url, err := url.Parse(importer.config.URL)
+	if err != nil {
+		return err
+	}
+
+	appender := importer.headBlock.Appender()
+	// todo: do 2 hr blocks here?
```

Comment:
Yes, you want to do this in no more than 2h chunks. Doing every step individually would be much slower than it needs to be.

Reply:
This is where `blocks.Writer` from https://github.com/prometheus/prometheus/pull/7586 might be helpful (: